src/impl_/extract_argument.rs

```diff
@@ -360,50 +334,77 @@ impl FunctionDescription {
         let varargs = V::handle_varargs_tuple(args, self)?;
 
         // Handle keyword arguments
-        let mut varkeywords = Default::default();
+        let mut varkeywords = K::Varkeywords::default();
         if let Some(kwargs) = kwargs {
-            let mut positional_only_keyword_arguments = Vec::new();
-            for (kwarg_name_py, value) in kwargs {
-                // All keyword arguments should be UTF8 strings, but we'll check, just in case.
-                if let Ok(kwarg_name) = kwarg_name_py.downcast::<PyString>()?.to_str() {
-                    // Try to place parameter in keyword only parameters
-                    if let Some(i) = self.find_keyword_parameter_in_keyword_only(kwarg_name) {
-                        if output[i + num_positional_parameters]
-                            .replace(value)
-                            .is_some()
-                        {
-                            return Err(self.multiple_values_for_argument(kwarg_name));
-                        }
-                        continue;
+            self.handle_kwargs::<K>(kwargs, &mut varkeywords, num_positional_parameters, output)?
+        }
+
+        // Once all inputs have been processed, check that all required arguments have been provided.
+
+        self.ensure_no_missing_required_positional_arguments(output, args.len())?;
+        self.ensure_no_missing_required_keyword_arguments(output)?;
+
+        Ok((varargs, varkeywords))
+    }
+
+    #[inline]
+    fn handle_kwargs<'py, K>(
+        &self,
+        kwargs: impl IntoIterator<Item = (&'py PyAny, &'py PyAny)>,
+        varkeywords: &mut K::Varkeywords,
+        num_positional_parameters: usize,
+        output: &mut [Option<&'py PyAny>],
+    ) -> PyResult<()>
+    where
+        K: VarkeywordsHandler<'py>,
+    {
+        debug_assert_eq!(
+            num_positional_parameters,
+            self.positional_parameter_names.len()
+        );
+        debug_assert_eq!(
+            output.len(),
+            num_positional_parameters + self.keyword_only_parameters.len()
+        );
+        let mut positional_only_keyword_arguments = Vec::new();
+        for (kwarg_name_py, value) in kwargs {
+            // All keyword arguments should be UTF8 strings, but we'll check, just in case.
+            if let Ok(kwarg_name) = kwarg_name_py.downcast::<PyString>()?.to_str() {
```

Comment:
This was not changed in the PR, but shouldn't we fail instead of silently ignoring non-UTF-8 keyword arguments?

Reply:
Non-UTF-8 keywords can't possibly be keywords to a `#[pyfunction]`, so we can just forward them to the varkeywords handler, which will either raise or add to `**kwargs` accordingly.

Reply:
I've pushed a comment to clarify this.

Reply:
I see, we do not actually skip them, just forward them instead.